helm/cluster-openstack/templates/list.yaml

```diff
@@ -0,0 +1,11 @@
+apiVersion: v1
+kind: List # Using a list to ensure templates/ClusterClass are created before the cluster itself
+items:
+- {{- include "cluster-class" . | indent 2 }}
+- {{- include "openstack-cluster-template" . | indent 2 }}
+- {{- include "kubeadm-control-plane-template" . | indent 2 }}
+{{- range .Values.nodeClasses }}
+- {{- include "kubeadm-config-template" (set (set (set . "Values" $.Values) "Chart" $.Chart) "Release" $.Release) | indent 2 }}
```

Comment:
There's something strange with scope when using nested templates (`{{ include }} inside another {{ include }}`). I couldn't get `_helpers.tpl` to work in the child templates without this hack. I could either pass in `{{- include "kubeadm-config-template" . }}` in which case `.Values`, `.Chart`, and `.Release` weren't available (even when prefixed with `$`). Or if I used `{{- include "kubeadm-config-template" $ }}` then the loop scope wasn't available. @giantswarm/team-honeybadger Let me know if you know any other solutions.

Reply:
This is really unfortunate. I am not certain how the scoping in helm works exactly. Maybe slightly nicer would be to use a `dict` instead of multiple `set` invocations:
https://helm.sh/docs/chart_template_guide/function_list/#dictionaries-and-dict-functions

Also this blog post I found: https://itnext.io/use-named-templates-like-functions-in-helm-charts-641fbcec38da

Reply:
Thanks for the resources. The `var_dump` function in the blog post was very useful. It showed me that `.Release` etc. weren't present inside the loop. I used a variable to capture the outer scope and then the `merge` function to merge the outer and inner scope. Feels a little better now.